lib/operations/create_index.js

```diff
@@ -1,212 +1,89 @@
-     * 11000 = 'DuplicateKey' (couldn't build unique index because of dupes)
-     * 11600 = 'InterruptedAtShutdown' (interrupted at shutdown)
-     * 197 = 'InvalidIndexSpecificationOption' (`_id` with `background: true`)
-     */
-    if (
-      err.code === 67 ||
-      err.code === 11000 ||
-      err.code === 85 ||
-      err.code === 86 ||
-      err.code === 11600 ||
-      err.code === 197
-    ) {
-      return handleCallback(callback, err, result);
-    }
-
-    // Create command
-    const doc = createCreateIndexCommand(db, name, fieldOrSpec, options);
-    // Set no key checking
-    finalOptions.checkKeys = false;
-    // Insert document
-    db.s.topology.insert(
-      db.s.namespace.withCollection(CONSTANTS.SYSTEM_INDEX_COLLECTION),
-      doc,
-      finalOptions,
-      (err, result) => {
-        if (callback == null) return;
-        if (err) return handleCallback(callback, err);
-        if (result == null) return handleCallback(callback, null, null);
-        if (result.result.writeErrors)
-          return handleCallback(callback, MongoError.create(result.result.writeErrors[0]), null);
-        handleCallback(callback, null, doc.name);
+    // Build the index
+    const indexParameters = parseIndexOptions(fieldOrSpec);
+    // Generate the index name
+    const indexName = typeof options.name === 'string' ? options.name : indexParameters.name;
+    // Set up the index
+    const indexesObject = { name: indexName, key: indexParameters.fieldHash };
+    // merge all the options
+    for (let optionName in options) {
+      if (!keysToOmit.has(optionName)) {
+        indexesObject[optionName] = options[optionName];
       }
-    );
-  });
-}
-
-/**
- * Create index using the createIndexes command.
- *
- * @param {Db} db The Db instance on which to execute the command.
- * @param {string} name Name of the collection to create the index on.
- * @param {(string|object)} fieldOrSpec Defines the index.
- * @param {Object} [options] Optional settings. See Db.prototype.createIndex for a list of options.
- * @param {Db~resultCallback} [callback] The command result callback.
- */
-function createIndexUsingCreateIndexes(db, name, fieldOrSpec, options, callback) {
-  // Build the index
-  const indexParameters = parseIndexOptions(fieldOrSpec);
-  // Generate the index name
-  const indexName = typeof options.name === 'string' ? options.name : indexParameters.name;
-  // Set up the index
-  const indexes = [{ name: indexName, key: indexParameters.fieldHash }];
-  // merge all the options
-  const keysToOmit = Object.keys(indexes[0]).concat([
-    'writeConcern',
-    'w',
-    'wtimeout',
-    'j',
-    'fsync',
-    'readPreference',
-    'session'
-  ]);
-
-  for (let optionName in options) {
-    if (keysToOmit.indexOf(optionName) === -1) {
-      indexes[0][optionName] = options[optionName];
     }
-  }
 
-  // Get capabilities
-  const capabilities = db.s.topology.capabilities();
+    this.indexName = indexName;
```

Comment:
I think this is not desirable to have here, but it is necessary if we move index construction into `_buildCommand`.

Reply:
can't you just get the value from the returned command?
```
const command = this._buildCommand();
const indexName = command.indexes[0].name;
```

Reply:
sure, but then I have to store the command twice since `CommandOperation.execute` does `command = this._buildCommand`.